### PR TITLE
[Ruby 3.0 support] Handle Kernel#clone with `freeze: true` argument

### PR DESCRIPTION
**Pull request by andrykonchin**
This PR addresses issues:

> [medium, java] Kernel#clone when called with the freeze: false keyword will call
#initialize_clone with the freeze: false keyword.
[Bug #14266]

and

> [medium, java] Kernel#clone when called with the freeze: true keyword will call
#initialize_clone with the freeze: true keyword, and will
return a frozen copy even if the receiver is unfrozen.
[Feature #16175]

Related issue - https://github.com/oracle/truffleruby/issues/2453

## Changes
- return a frozen copy when `freeze: true`
- handle `freeze: nil`
- pass `freeze:` to `#initialize_clone`

### CODE REVIEW

File: src/main/java/org/truffleruby/core/kernel/KernelNodes.java [hunk 1]
```diff
@@ -576,9 +577,31 @@ protected RubyDynamicObject clone(RubyDynamicObject object, boolean freeze,
...
+                final Object[] newStore = PackedHashStoreLibrary.createStore();
+                final int hashed = hashNode.execute(key);
+                PackedHashStoreLibrary.setHashedKeyValue(newStore, 0, hashed, key, value);
+
+                final RubyHash hash = new RubyHash(
+                        coreLibrary().hashClass,
+                        getLanguage().hashShape,
+                        getContext(),
+                        newStore,
+                        1);
```
Comment (andrykonchin) [inline src/main/java/org/truffleruby/core/kernel/KernelNodes.java:596]: I am not familiar with the Rope concept and how to build it properly for building a Symbol. So any example or clarification will be very helpful 🙇 .

```ruby
PackedHashStoreLibrary.setHashedKeyValue(newStore, 0, hashed, key, value);
```
I assume here that `0` is an index of the first element in a 0-based array.

```ruby
new RubyHash(
    coreLibrary().hashClass,
    getLanguage().hashShape,
    getContext(),
    newStore,
    1);
```

Here I suppose `1` means how many elements are in the Hash

Also, I suppose I work with Symbols in the wrong way - a new symbol should be probably registered in some repository but not just created. Please advise me how to do it properly 🙇.

**Comment by andrykonchin** [comment]
I would appreciate any advice regarding failed specs:

```
1)
Kernel#clone with freeze: true calls #initialize_clone with kwargs freeze: true FAILED
Expected [#<KernelSpecs::CloneFreeze:0x558>, {[:freeze, true]=>{}}] == [#<KernelSpecs::CloneFreeze:0x558>, {[:freeze, true]=>{}}]
to be truthy but was false
...spec/ruby/core/kernel/clone_spec.rb:97:in `block (4 levels) in <top (required)>'
...spec/ruby/core/kernel/clone_spec.rb:5:in `<top (required)>'

2)
Kernel#clone with freeze: false calls #initialize_clone with kwargs freeze: false FAILED
Expected [#<KernelSpecs::CloneFreeze:0x638>, {[:freeze, false]=>{}}] == [#<KernelSpecs::CloneFreeze:0x638>, {[:freeze, false]=>{}}]
to be truthy but was false
...spec/ruby/core/kernel/clone_spec.rb:129:in `block (4 levels) in <top (required)>'
...spec/ruby/core/kernel/clone_spec.rb:5:in `<top (required)>'
```

So actually `{[:freeze, true]=>{}}` doesn't equal `{[:freeze, true]=>{}}`.

I suppose there is an issue with string representing of RubyHash, actually it's a correct Hash `{ freeze: true }`.

And finally a key `:freeze` in recorded arguments doesn't equal a Symbol literal `:freeze` in the test case. I suppose it's related to how the key was created in the `clone` Java method.

File: src/main/java/org/truffleruby/core/kernel/KernelNodes.java [hunk 2]
```diff
@@ -590,63 +586,95 @@ protected RubyDynamicObject clone(RubyDynamicObject object, boolean freeze,
...
+        private boolean toForceFreezing(Object freeze) {
+           if (freeze instanceof Nil) {
+               return false;
+           }
+
+           if (freeze instanceof Boolean) {
+               return (boolean) freeze;
+           }
+
+           throw new RaiseException(getContext(), coreExceptions().argumentError(
```
[the remark this answers is not in the document]
Comment (andrykonchin) [reply src/main/java/org/truffleruby/core/kernel/KernelNodes.java]: I will change the wording but will look at it a bit later.

File: src/main/java/org/truffleruby/core/kernel/KernelNodes.java (hunk 1 above)
Comment (eregon) [reply src/main/java/org/truffleruby/core/kernel/KernelNodes.java:596]: Yes, that's all correct and looks good.

File: src/main/java/org/truffleruby/core/kernel/KernelNodes.java [hunk 3]
```diff
@@ -576,9 +577,31 @@ protected RubyDynamicObject clone(RubyDynamicObject object, boolean freeze,
                 newObjectMetaClass.fields.initCopy(selfMetaClass);
             }
 
-            initializeCloneNode.call(newObject, "initialize_clone", object);
+            // pass :freeze keyword argument to #initialize_clone
+            if (toForceFreezing(freeze) || toForceUnfreezing(freeze)) {
+                final String string = "freeze";
+                final LeafRope rope = RopeOperations.encodeAscii(string, USASCIIEncoding.INSTANCE);
+                final RubySymbol key = new RubySymbol(string, rope, Encodings.US_ASCII);
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/kernel/KernelNodes.java:584]: The RubySymbol should be package-private so it's clear it shouldn't be used directly.
There is already `coreSymbols().FREEZE` which you can use :)

File: src/main/java/org/truffleruby/core/kernel/KernelNodes.java [hunk 4]
```diff
@@ -576,9 +577,31 @@ protected RubyDynamicObject clone(RubyDynamicObject object, boolean freeze,
                 newObjectMetaClass.fields.initCopy(selfMetaClass);
             }
 
-            initializeCloneNode.call(newObject, "initialize_clone", object);
+            // pass :freeze keyword argument to #initialize_clone
+            if (toForceFreezing(freeze) || toForceUnfreezing(freeze)) {
```
[the remark this answers is not in the document]
Comment (eregon) [reply src/main/java/org/truffleruby/core/kernel/KernelNodes.java:581]: Actually checking `freeze` is `nil` or a boolean would probably be nicest in its own Primitive or node with `@CreateCast`, I'll do that.
You can for instance move that check in the Ruby definition and I'll optimize it later.

File: src/main/java/org/truffleruby/core/kernel/KernelNodes.java (hunk 3 above)
Comment (andrykonchin) [reply src/main/java/org/truffleruby/core/kernel/KernelNodes.java:584]: Thank you. It fixed the failed specs as well 🎆

File: src/main/java/org/truffleruby/core/kernel/KernelNodes.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (andrykonchin) [reply src/main/java/org/truffleruby/core/kernel/KernelNodes.java:581]: Moved to the Ruby code.

**Comment by andrykonchin** [comment]
Done. Please review.

**Review by eregon — APPROVED** [review]
Thank you for the great work on this PR!

**Comment by andrykonchin** [comment]
Thank you for your help 🙇.

I will squash commits and rebase on the master branch soon.

**Comment by eregon** [comment]
I pushed a couple commits on top, I'll do the rebase now.

**Comment by eregon** [comment]
I squashed the second commit in the first one to make it easier to rebase, should be all good now.

**Pull request merged**